third_party/intel/backend/compiler.py

```diff
@@ -421,42 +424,57 @@ def make_zebin(cls, src, metadata, options):
-                        warning: kernel matmul_kernel  compiled SIMD16 allocated 128 regs and spilled around 217
-                    is "spilled" enough for now?
-                    """
-                    metadata["build_flags"] += " -cl-intel-256-GRF-per-thread"
-                    # re-run with new build flags
-                    ocloc_cmd[-1] = metadata["build_flags"] + shader_dump_opt
+                try:
                     subprocess.check_output(ocloc_cmd, stderr=subprocess.STDOUT, text=True)
-            except subprocess.CalledProcessError as e:
-                if e.returncode == 255:
-                    error = 'Internal Triton ZEBIN codegen error'
-                elif e.returncode == 128 + signal.SIGSEGV:
-                    error = '`ocloc` raised SIGSEGV'
-                else:
-                    error = f'`ocloc` failed with error code {e.returncode}'
-
-                raise RuntimeError(f'{error}\n'
-                                   f'`ocloc` stderr:\n{e.output}\n'
-                                   f'Repro command: {ocloc_cmd}\n') from e
-
-            with open(fbin, 'rb') as f:
-                zebin = f.read()
+                except subprocess.CalledProcessError as e:
+                    if e.returncode == 255:
+                        error = 'Internal Triton ZEBIN codegen error'
+                    elif e.returncode == 128 + signal.SIGSEGV:
+                        error = '`ocloc` raised SIGSEGV'
+                    else:
+                        error = f'`ocloc` failed with error code {e.returncode}'
+
+                    raise RuntimeError(f'{error}\n'
+                                       f'`ocloc` stderr:\n{e.output}\n'
+                                       f'Repro command: {ocloc_cmd}\n') from e
+
+                with open(fbin, 'rb') as f:
+                    zebin = f.read()
+                    spill_size = 0
+                    if extract_spill_size:
+                        elf = ELFFile(io.BytesIO(zebin))
+                        zeinfo = elf.get_section_by_name(".ze_info")
+                        if zeinfo is None:
+                            raise RuntimeError('Internal Triton ZEBIN codegen error:'
+                                               'Section .ze_info not found in zebin')
+                        data = zeinfo.data()
+                        text = data.decode('utf-8')
+                        match = re.search(r'spill_size\s*[:=]\s*(\d+)', text)
+                        if match:
+                            spill_size = int(match.group(1))
+            return zebin, spill_size
+
+        auto_grf = options.grf_mode == 'default'
+        zebin, spill_size = build_zebin(metadata["build_flags"], extract_spill_size=auto_grf)
+
+        # The threshold of 1000 for spill_size is chosen based on empirical observations
+        # and aligned with triton/backends/intel/driver.c
+        if auto_grf and spill_size > 1000:
```

Comment:
Since we have both options:
```python
        elif options.grf_mode == 'auto':
            metadata["build_flags"] += " -cl-intel-enable-auto-large-GRF-mode"
        elif options.grf_mode != 'default':
            raise RuntimeError(f"Unknown grf_mode: {options.grf_mode}")
```

this could be clearer.

```suggestion
        default_grf = options.grf_mode == 'default'
        zebin, spill_size = build_zebin(metadata["build_flags"], extract_spill_size=default_grf)

        # The threshold of 1000 for spill_size is chosen based on empirical observations
        # and aligned with triton/backends/intel/driver.c
        if default_grf and spill_size > 1000:
```